### PR TITLE
v2.0.0 - Role Base

### PR DESCRIPTION
**Pull request by Tommy-XL**
### Based on: [EndlessHostRoles](https://github.com/Gurge44/EndlessHostRoles)

This is done to make the code shorter and easier

## Estimated completion time in 3 or 4 weeks from 03/03/24
~~It may be completed earlier than planned, in the worst case it will take more than 4 weeks to complete~~
Changes: this will take longer than planned, but this code will be used in `v2.0.0`

### 09/04/24 - all progress done (need test)

Progress: 
- [x] `Impostor ghost`: 100% done (Total 2)
- [x] `Crewmate ghost`: 100% done (Total 2)
- [x] `Impostors`: 100% done (Total 72) (`Evil Mini` done)
- [x] `Crewmates`: 100% done (Total 70) (`Nice Mini` done)
- [x] `Neutrals`: 100% done (Total 69) (Done in: #767)
- [x] Complete Setup for all roles (check this https://github.com/0xDrMoe/TownofHost-Enhanced/pull/693#issuecomment-2002530665)
- [x] Replace `List` with `HashSet` where is possible
- [x] Make `List`, `HashSet`, `Dictionary` as `readonly` and in `Init()` use `Clear()` instead of `= [];` where is possible
- [x] Rework the code execution algorithm in `CheckMurderPatch`
- [x] `RpcMurderPlayerV3` => `RpcMurderPlayer` 
- [x] Add `RpcCheckAndMurder` & `CheckForInvalidMurdering` in some places of the code
- [x] Maybe need making changes to `MurderPlayerPatch` (Done in: #767)
- [x] Sunnyboy is partially broken
- [x] Move execution `OnTaskComplete()` code
- [x] Fix and return ghost roles
- [x] Recode `CheckShapeshiftPatch` 
- [x] Change all `OnShapeshift()`
- [x] Some fix in `KnowRole`

### CODE REVIEW

**Comment by NikoCat233** [comment]
How about ScientistTOHE and EngineerTOHE

**Comment by NikoCat233** [comment]
Gonna help with crew roles

**Comment by Ultradragon005** [comment]
image
image

Toughts? for getVnRole obviously.

**Comment by Tommy-XL** [comment]
> Toughts? for getVnRole obviously.

Nice

**Comment by Ultradragon005** [comment]
alrighty, I'll do it for all the roles that have now been done.

**Comment by Ultradragon005** [comment]
I did it #710

All of the currenct one (excepct cursedwolf & convict) have been added to it.

File: Roles/Core/CustomRoleManager.cs [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+
+        // ==== Crewmates ====
+        CustomRoles.Engineer => new VanillaRole(),
+        CustomRoles.GuardianAngel => new VanillaRole(),
+        CustomRoles.Scientist => new VanillaRole(),
+        CustomRoles.CrewmateTOHE => new VanillaRole(),
+        CustomRoles.EngineerTOHE => new VanillaRole(),
+        CustomRoles.GuardianAngelTOHE => new VanillaRole(),
+        CustomRoles.ScientistTOHE => new VanillaRole(),
+        //CustomRoles.Addict => new Addict(),
```
Comment (Ultradragon005) [inline Roles/Core/CustomRoleManager.cs:114]: forgot to uncomment here

Comment (Tommy-XL) [reply Roles/Core/CustomRoleManager.cs]: Fixed

**Comment by Tommy-XL** [comment]
Progress:
1. Need to complete all neutrals
2. Sunnyboy is partially broken
3. Complete Setup for all roles as in the photo

image

**Comment by Tommy-XL** [comment]
4. Replace `List` with `HashSet` where possible
5. Make `List`, `HashSet`, `Dictionary` as `readonly` and in `Init()` use `Clear()` instead of `= [];` where possible

**Comment by Ultradragon005** [comment]
> Progress:
> 
> 1. Need to complete all neutrals
> 2. Sunnyboy is partially broken
> 3. Complete Setup for all roles as in the photo

Ngl, I figured out you can actually do the `bool On` thingy. Tho it unneccesarily gets set to true again if someone gets it midgame. which is why I prefer HashSet PlayerIdList;

**Comment by Tommy-XL** [comment]
6. Recode `CheckShapeshiftPatch`

**Comment by Ultradragon005** [comment]
7. Need Complete All Ghost Roles

**Comment by Ultradragon005** [comment]
RecieveRpc is also possible. But for that we will need to make Dictionary<CustomRole, RoleBase> EnabledRoles; (Ig FirstOrDeafult() `playerstate` or new RoleBase(); also technicaly work, but there's other implications to enabled roles such as onplayerleftpatch)

ANd I think I'm gonan do that later, cause it's gonne remove so so many customrpc's

**Comment by Tommy-XL** [comment]
8. Move execution `OnTaskComplete` code

**Comment by Tommy-XL** [comment]
> RecieveRpc is also possible. But for that we will need to make Dictionary<CustomRole, RoleBase> EnabledRoles; (Ig FirstOrDeafult() `playerstate` or new RoleBase(); also technicaly work, but there's other implications to enabled roles such as onplayerleftpatch)
> 
> ANd I think I'm gonan do that later, cause it's gonne remove so so many customrpc's

NikoCat planned to do this

**Comment by Ultradragon005** [comment]
> > RecieveRpc is also possible. But for that we will need to make Dictionary<CustomRole, RoleBase> EnabledRoles; (Ig FirstOrDeafult() `playerstate` or new RoleBase(); also technicaly work, but there's other implications to enabled roles such as onplayerleftpatch)
> > ANd I think I'm gonan do that later, cause it's gonne remove so so many customrpc's
> 
> NikoCat planned to do this

oh okay

**Comment by Tommy-XL** [comment]
after Role Base i think

**Comment by Tommy-XL** [comment]
9. Rework the code execution algorithm in `CheckMurderPatch`

10. It might also be worth making changes to `MurderPlayerPatch`

**Comment by Ultradragon005** [comment]
> ### Thanks [TOHE+](https://github.com/Gurge44/TOHE_PLUS)
> This is done to make the code shorter and easier
> 
> ## Estimated completion time in 3 or 4 weeks from 03/03/24
> ~It may be completed earlier than planned, in the worst case it will take more than 4 weeks to complete~ Changes: this will take longer than planned, but this code will be used in `v2.0.0`
> 
> Progress:
> 
> * [ ]  `Impostor ghost`: 0 done (Total 2)
> * [ ]  `Crewmate ghost`: 0 done (Total 2)
> * [x]  `Impostors`: 100% done (Total 72) (`Evil Mini` done)
> * [x]  `Crewmates`: 100% done (Total 70) (`Nice Mini` done)
> * [ ]  `Neutrals`: 53 done (Total 69)
> * [ ]  Sunnyboy is partially broken
> * [ ]  Complete Setup for all roles (check this [Role Base (based on TOHE+) planned used it in v2.0.0 #693 (comment)](https://github.com/0xDrMoe/TownofHost-Enhanced/pull/693#issuecomment-2002530665))
> * [ ]  Replace `List` with `HashSet` where is possible
> * [ ]  Make `List`, `HashSet`, `Dictionary` as `readonly` and in `Init()` use `Clear()` instead of `= [];` where is possible
> * [ ]  Recode `CheckShapeshiftPatch` and change `OnShapeshift()`
> * [ ]  Move execution `OnTaskComplete()` code
> * [ ]  Rework the code execution algorithm in `CheckMurderPatch`
> * [ ]  Maybe need making changes to `MurderPlayerPatch`

should probably also change as many List/Dictionary/Hashset etc.. .Count > 0; => .Any(); Because after some digging I found out Any() checks for the first iteration of a value, if none returns false otherwise true; While .Count >0; first counts all the itterations in the list and then compares to 0;

**Comment by Tommy-XL** [comment]
So much to do...
But not today, xd

**Comment by Tommy-XL** [comment]
`ReportDeadBodyPatch` only fires once, meaning if you cancel the report, clients will no longer be able to call `PlayerControl.ReportDeadBody`

Possible fix:
Remember the player who tried to report the dead body and if he can be reported again then you need to check the position of the player and the dead body, if he is alive and comes close to the dead body then immediately call `RpcStartMeeting` via `FixedUpdate`

I'm thinking of doing this after Role Base is tested in dev 1, 2, 3 etc.

**Comment by Tommy-XL** [comment]
For modded just can do `Dead Body.Reported = false;`

**Comment by NikoCat233** [comment]
> `ReportDeadBodyPatch` only fires once, meaning if cancel the report, clients will no longer be able to call `PlayerControl.ReportDeadBody`
> 
> Possible fix: Remember the player who tried to report the dead body and if dead body can be reported again then need to check the position of the player and the dead body, if player is alive and comes close to the dead body then immediately call `RpcStartMeeting` via `FixedUpdate`
> 
> I'm thinking of doing this after Role Base is tested in dev 1, 2, 3 etc.

If the bodies' owner is still in game, u can litterly send another rpcmurderplayer to the reporter client to create a new body for it to report i think

**Comment by Tommy-XL** [comment]
I think it’s easier and better to implement my option, because the ghost sometimes needs to constantly teleport to the body and this may not work when the ghost automatically follows some player

**Comment by NikoCat233** [comment]
> I think it’s easier and better to implement my option, because the ghost sometimes needs to constantly teleport to the body and this may not work when the ghost automatically follows some player

No u only send murder calls to the reporter client
only reporter will see ghost teleported u know

**Comment by Tommy-XL** [comment]
I think this PR will be merged with the main one after the first version of Canary is released

**Pull request merged**